File: Learning_Japan_project_fe/learning-app-fe/src/services/userService.ts
import axios from "axios";
import { axiosClient, HttpClientError } from "@/lib/axios";
import { getEndpoint } from "@/config/api";
import { ApiResponse } from "./api-types"; // <-- dùng từ file api-types
import { API_ENDPOINTS } from "@/config/api";
import { JLPTLevel } from "@/enums/JLPTLevel";
import { http } from "@/lib/http";
export interface RegisterRequest {
  fullName: string;
  email: string;
  password: string;
}

export interface RegisterResult {
  id: string;
  fullName: string;
  email: string;
  createdAt: string;
}

export interface UserProfileResponse {
  fullName: string;
  email: string;
  createdAt: string;
  avatarUrl: string;
  level: JLPTLevel;
  avatar?: string;
  isPremium?: boolean;
  roles?: string[];
}

export interface UserResponse {
  id: string;
  email: string;
  fullName: string; // Backend là fullName, không phải name
  avatarUrl?: string;
  enabled: boolean;
  level: JLPTLevel;

  stage: string; // "Junbi"
  processPercent: number; // 60
  isPremium: boolean; // true/false

  createdAt: string;
}

export interface UserResponseManager {
  id: string;
  email: string;
  fullName: string; // Backend là fullName, không phải name
  avatarUrl?: string;
  enabled: boolean;
  role: string[]; // Ví dụ: ["USER"], ["ADMIN"]

  // Các field học tập mới thêm
  level: string; // "N5"
  stage: string; // "Junbi"
  processPercent: number; // 60
  isPremium: boolean; // true/false

  createdAt: string;
}
export interface UserChatResponse {
  id: string;
  fullName: string;
  email: string;
  avatarUrl?: string; // Một số chỗ dùng avatarUrl
  avatar?: string;    // Một số chỗ dùng avatar
  level?: string;
  isPremium?: boolean;
}
export interface UpdateUserRequest {
  fullName: string; // Backend là fullName, không phải name
  level: JLPTLevel;

}
export interface PageResponse<T> {
  page: number;
  totalPages: number;
  size: number;
  totalElements: number;
  data: T[]; // Danh sách user nằm ở đây
}

export interface UserStatsResponse {
  total: number;
  active: number;
  banned: number;
}

export const userService = {
  getProfile(): Promise<UserProfileResponse> {
    return http.get(API_ENDPOINTS.USER.PROFILE); // token đi kèm trong header
  },

  updateProfile(data: UpdateUserRequest): Promise<UserProfileResponse> {
    return http.put(API_ENDPOINTS.USER.UPDATE_PROFILE, data);
  },

  uploadAvatar(file: File): Promise<void> {
    const formData = new FormData();
    formData.append("file", file);
    return http.put(API_ENDPOINTS.USER.UPLOAD_AVATAR, formData);
  },

  changePassword(currentPassword: string, newPassword: string): Promise<void> {
    return http.post(API_ENDPOINTS.USER.CHANGE_PASSWORD, {
      currentPassword,
      newPassword,
    });
  },

  // lấy tất cả user cho admin
  getAllUsers(
    page: number,
    size: number,
    search?: string
  ): Promise<PageResponse<UserResponse>> {
    const params: Record<string, string | number> = { page, size };
    if (search) {
      params.search = search;
    }
    return http.get(API_ENDPOINTS.USER.ALL_USERS, { params });
  },

  // lấy chi tiết user cho admin
  getAllUsersManager(
    page: number,
    size: number,
    search?: string
  ): Promise<PageResponse<UserResponseManager>> {
    const params: Record<string, string | number> = { page, size };
    if (search) {
      params.search = search;
    }
    return http.get(API_ENDPOINTS.ADMIN.ALL_USERS_MANAGER, { params });
  },

  /**
   * Ban (Khóa) tài khoản user dựa trên email
   * Endpoint: POST /api/admin/users/{email}/ban
   */
  banUser(email: string): Promise<void> {
    // Lưu ý: Đảm bảo đường dẫn khớp với Controller Backend bạn vừa viết
    // Nếu bạn chưa cấu hình trong API_ENDPOINTS, hãy viết trực tiếp string như dưới:
    return http.post(API_ENDPOINTS.ADMIN.BAN_USER(email));
  },

  /**
   * Unban (Mở khóa) tài khoản user dựa trên email
   * Endpoint: POST /api/admin/users/{email}/unban
   */
  unbanUser(email: string): Promise<void> {
    return http.post(API_ENDPOINTS.ADMIN.UNBAN_USER(email));
  },

  deleteUserAccount(email: string): Promise<void> {
    return http.delete(API_ENDPOINTS.ADMIN.DELETE_USER, {
      params: { email },
    });
  },

  deleteMultipleUserAccounts(emails: string[]): Promise<void> {
    return http.delete(API_ENDPOINTS.ADMIN.DELETE_USERS, {
      data: { emails },
    });
  },
  searchUsers(keyword: string): Promise<UserChatResponse[]> {
    return http.get(API_ENDPOINTS.USER.SEARCH, {
      params: { keyword },
    });
  },

  getUserStatistics(): Promise<UserStatsResponse> {
    return http.get(API_ENDPOINTS.ADMIN.USER_STATISTICS);
  },
  getUserById(id: string): Promise<UserChatResponse> {
    return http.get<UserChatResponse>(API_ENDPOINTS.USER.GET_USER_BY_ID(id));
  },
};

export const register = async (
  data: RegisterRequest
): Promise<RegisterResult> => {
  try {
    const endpoint = getEndpoint("REGISTER");

    // Gọi API, kiểu ApiResponse<RegisterResult>
    const response = await axiosClient.post<ApiResponse<RegisterResult>>(
      endpoint,
      data,
      { headers: { Authorization: undefined } }
    );

    const resData = response.data;

    // Check success từ backend
    if (!resData.success) {
      throw new Error(resData.message || "Đăng ký thất bại");
    }

    return resData.result;
  } catch (error: unknown) {
    // Xử lý lỗi
    if (error instanceof HttpClientError) {
      throw new Error(error.message || "Đăng ký thất bại");
    }

    if (axios.isAxiosError(error)) {
      const msg =
        (error.response?.data as { message?: string })?.message ||
        error.message;
      throw new Error(msg);
    }

    if (error instanceof Error) {
      throw new Error(error.message);
    }

    throw new Error("Có lỗi xảy ra");
  }
};

File: Learning_Japan_project/learningApp/src/main/java/com/example/learningApp/service/user/UserService.java
package com.example.learningApp.service.user;

import com.example.learningApp.common.PageResponse;
import com.example.learningApp.dto.request.user.DeleteUsersRequest;
import com.example.learningApp.dto.request.user.UpdateUserRequest;
import com.example.learningApp.dto.response.user.UserChatResponse;
import com.example.learningApp.dto.response.user.UserForAdminResponse;
import com.example.learningApp.dto.response.user.UserResponse;
import com.example.learningApp.dto.response.user.UserStatsResponse;
import com.example.learningApp.entity.Role;
import com.example.learningApp.entity.User;
import com.example.learningApp.mapper.UserMapper;
import com.example.learningApp.repository.UserLearningProgressRepository;
import com.example.learningApp.repository.UserRepository;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import lombok.experimental.NonFinal;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;
import software.amazon.awssdk.services.cognitoidentityprovider.CognitoIdentityProviderClient;
import software.amazon.awssdk.services.cognitoidentityprovider.model.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import static com.example.learningApp.configuration.CognitoSecretHashUtil.calculateSecretHash;

@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
@Slf4j
public class UserService {
    private final UserMapper userMapper;
    @NonFinal
    @Value("${aws.cognito.client-id}")
    String clientId;
    @NonFinal
    @Value("${aws.cognito.client-secret}")
    String clientSecret;
    @NonFinal
    @Value("${aws.cognito.user-pool-id}")
    String userPoolId;

    CognitoIdentityProviderClient cognitoClient;
    UserRepository userRepository;
    UserLearningProgressRepository userLearningProgressRepository;

    public void changePassword(String accessToken, String oldPassword, String newPassword) {
        try {
            ChangePasswordRequest changePasswordRequest = ChangePasswordRequest.builder()
                    .accessToken(accessToken)
                    .previousPassword(oldPassword)
                    .proposedPassword(newPassword)
                    .build();

            // Call Cognito to change the password
            cognitoClient.changePassword(changePasswordRequest);
        } catch (NotAuthorizedException e) {
            throw new RuntimeException("Old password is incorrect");
        } catch (InvalidPasswordException e) {
            throw new RuntimeException("New password does not meet security policy");
        } catch (Exception ex) {
            throw new RuntimeException("Change password failed", ex);
        }
    }

    @Transactional
    public UserResponse updateMyProfile(UpdateUserRequest request) {

        var context = SecurityContextHolder.getContext();
        String userId = context.getAuthentication().getName();

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (request.getFullName() != null) {
            user.setFullName(request.getFullName());
        }

        if (request.getLevel() != null) {
            user.setLevel(request.getLevel());
        }

        userRepository.save(user);

        return userMapper.toUserResponse(user);
    }

    public void forgetPassword(String email) {
        // Implementation for forget password
        try {
            ForgotPasswordRequest forgotPasswordRequest = ForgotPasswordRequest.builder()
                    .clientId(clientId)
                    .username(email)
                    .secretHash(calculateSecretHash(email, clientId, clientSecret))
                    .build();
            cognitoClient.forgotPassword(forgotPasswordRequest);
        } catch (UserNotFoundException e) {
            log.warn("Forgot password requested for non-existing user");
        } catch (Exception e) {
            throw new RuntimeException("Forget password failed", e);
        }
    }

    public void confirmForgotPassword(String email, String confirmationCode, String newPassword) {
        try {
            ConfirmForgotPasswordRequest confirmForgotPasswordRequest = ConfirmForgotPasswordRequest.builder()
                    .clientId(clientId)
                    .username(email)
                    .confirmationCode(confirmationCode)
                    .password(newPassword)
                    .secretHash(calculateSecretHash(email, clientId, clientSecret))
                    .build();
            cognitoClient.confirmForgotPassword(confirmForgotPasswordRequest);
        } catch (CodeMismatchException e) {
            throw new RuntimeException("Invalid confirmation code");
        } catch (ExpiredCodeException e) {
            throw new RuntimeException("Confirmation code expired");
        } catch (InvalidPasswordException e) {
            throw new RuntimeException("Password does not meet policy");
        } catch (Exception e) {
            throw new RuntimeException("Confirm forgot password failed", e);
        }
    }

    public PageResponse<UserForAdminResponse> getAllUsersManager(int page, int size, String search) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<User> userPage;
        if (search == null) {
            userPage = userRepository.findAll(pageable);
        } else {
            userPage = userRepository.findByEmailContainingIgnoreCaseOrFullNameContainingIgnoreCase(
                    search, search, pageable);
        }

        List<UserForAdminResponse> userForAdminResponse = userPage.getContent().stream().map(this::mapToUserResponse)
                .toList();
        return PageResponse.<UserForAdminResponse>builder()
                .page(page)
                .totalPages(userPage.getTotalPages())
                .size(userPage.getSize())
                .totalElements(userPage.getTotalElements())
                .data(userForAdminResponse)
                .build();
    }

    public PageResponse<UserForAdminResponse> getAllUsers(int page, int size, String search) {
        Pageable pageable = PageRequest.of(page - 1, size, Sort.by("createdAt").descending());

        // 1. Tìm kiếm và phân trang User
        Page<User> userPage;
        if (search == null) {
            userPage = userRepository.findAll(pageable);
        } else {
            userPage = userRepository.findByEmailContainingIgnoreCaseOrFullNameContainingIgnoreCase(
                    search, search, pageable);
        }

        // 2. Map sang UserResponse và ĐIỀN THÊM DỮ LIỆU HỌC TẬP
        List<UserForAdminResponse> userForAdminResponse = userPage.getContent().stream().map(user -> {
            // Map cơ bản (id, email, name...)
            UserForAdminResponse response = mapToUserResponse(user);

            // Lấy progress mới nhất của user này
            var progress = userLearningProgressRepository.findFirstByUserIdOrderByLastExamAtDesc(user.getId());

            if (progress != null) {
                response.setLevel(progress.getLevel()); // Set Level

                // Tính % hoàn thành (ví dụ đơn giản)
                int percent = (progress.getTotalQuestionsDone() == 0) ? 0
                        : (int) ((double) progress.getCorrectQuestions() / progress.getTotalQuestionsDone() * 100);
                response.setProcessPercent(percent);

                // Set Stage (Ví dụ: Nếu làm > 10 bài thi là giai đoạn Exam, ngược lại là Junbi)
                response.setStage(progress.getTotalExamsTaken() > 10 ? "Exam" : "Junbi");
            } else {
                // Mặc định cho user mới chưa học gì
                response.setLevel("N5");
                response.setProcessPercent(0);
                response.setStage("Newbie");
            }

            // Set Premium (giả sử User entity có field này hoặc check role)
            // response.setPremium(user.getRoles().stream().anyMatch(r ->
            // r.getName().equals("ROLE_PREMIUM")));

            return response;
        }).collect(Collectors.toList());

        return PageResponse.<UserForAdminResponse>builder()
                .page(page)
                .totalPages(userPage.getTotalPages())
                .size(userPage.getSize())
                .totalElements(userPage.getTotalElements())
                .data(userForAdminResponse)
                .build();
    }

    // Helper map cơ bản
    private UserForAdminResponse mapToUserResponse(User user) {
        return UserForAdminResponse.builder()
                .id(user.getId())
                .email(user.getEmail())
                .fullName(user.getFullName())
                .enabled(user.getEnabled())
                .role(String.valueOf(user.getRoles().stream().map(Role::getRoleName).collect(Collectors.toList())))
                .avatarUrl(user.getAvatarUrl())
                .createdAt(user.getCreatedAt())
                .updatedAt(user.getUpdatedAt())
                .build();
    }

    @Transactional
    public void banUser(String email) {
        User user = userRepository.findByEmail(email).orElseThrow(() -> new RuntimeException("User not found"));
        user.setEnabled(false);
        userRepository.save(user);

        try {
            // Disable user in Cognito
            cognitoClient.adminDisableUser(AdminDisableUserRequest.builder()
                    .userPoolId(userPoolId)
                    .username(email)
                    .build());
        } catch (Exception e) {
            throw new RuntimeException("Ban user failed", e);
        }
    }

    @Transactional
    public void unbanUser(String email) {
        User user = userRepository.findByEmail(email).orElseThrow(() -> new RuntimeException("User not found"));
        user.setEnabled(true);
        userRepository.save(user);

        try {
            // Enable user in Cognito
            cognitoClient.adminEnableUser(AdminEnableUserRequest.builder()
                    .userPoolId(userPoolId)
                    .username(email)
                    .build());
        } catch (Exception e) {
            throw new RuntimeException("Unban user failed", e);
        }
    }

    public void deleteUser(String email) {
        User user = userRepository.findByEmail(email).orElseThrow(() -> new RuntimeException("User not found"));
        userRepository.delete(user);

        try {
            // Xoá user trong Cognito
            cognitoClient.adminDeleteUser(AdminDeleteUserRequest.builder()
                    .userPoolId(userPoolId)
                    .username(email)
                    .build());
        } catch (Exception e) {
            throw new RuntimeException("Delete user failed", e);
        }
    }

    public void deleteUsers(DeleteUsersRequest request) {
        for (String email : request.getEmails()) {
            try {
                // Tái sử dụng hàm delete đơn lẻ ở trên
                deleteUser(email);
            } catch (Exception e) {
                // Log lỗi nếu cần, nhưng tiếp tục xóa người tiếp theo
                System.err.println("Lỗi khi xóa user " + email + ": " + e.getMessage());
                // Tùy nghiệp vụ: Bạn có thể throw lỗi để dừng hẳn hoặc bỏ qua để xóa tiếp
            }
        }
    }

    public UserChatResponse getUserById(String userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        return UserChatResponse.builder()
                .id(user.getId())
                .fullName(user.getFullName())
                .avatarUrl(user.getAvatarUrl())
                .build();
    }

    // ============== SEARCH USER FOR CHAT =================
    public List<UserChatResponse> searchUsersForChat(String keyword) {

        var context = SecurityContextHolder.getContext();
        String currentUserId = context.getAuthentication().getName();

        Pageable pageable = PageRequest.of(0, 20); // giới hạn 20 user

        List<User> users = userRepository
                .findByFullNameContainingIgnoreCaseOrEmailContainingIgnoreCaseAndEnabledTrue(
                        keyword, keyword, pageable);

        return users.stream()
                .filter(user -> !user.getId().equals(currentUserId)) // không hiện chính mình
                .map(user -> UserChatResponse.builder()
                        .id(user.getId())
                        .fullName(user.getFullName())
                        .avatarUrl(user.getAvatarUrl())
                        .build())
                .toList();
    }

    // UserService.java
    public UserStatsResponse getUserStats() {
        Object result = userRepository.getUserStatisticsRaw();

        // Vì query aggregate trả về 1 dòng duy nhất nên ta ép kiểu sang Object[]
        if (result instanceof Object[] row) {

            // Cách an toàn nhất là ép về Number rồi lấy longValue().
            long total = ((Number) row[0]).longValue();
            long active = ((Number) row[1]).longValue();
            long banned = ((Number) row[2]).longValue();

            return new UserStatsResponse(total, active, banned);
        }

        return new UserStatsResponse(0, 0, 0);
    }

    // ============== USER ==================
    public UserResponse getMyInfo() {
        var context = SecurityContextHolder.getContext();
        String userId = context.getAuthentication().getName();
        User user = userRepository.findById(userId).orElseThrow();
        return userMapper.toUserResponse(user);
    }

    public void updateAvatarUrl(String url) {

        // lấy userid từ token
        var context = SecurityContextHolder.getContext();
        String userId = context.getAuthentication().getName();

        // tìm user
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        // cập nhật avatar
        user.setAvatarUrl(url);

        // lưu lại
        userRepository.save(user);
    }

}
